### DISCUSSION TITLE
Build failing on Mac (ld: library not found)

### DISCUSSION THREAD

**Discussion by ln53** [idaholab/moose#25652] [Q&A Getting Started] [answered]
Hi, I'm trying to install MOOSE locally on my laptop. I've followed the installation instructions, and the build issues section on the troubleshooting page, but the problem persists.

A whole load of `Compiling C++` is done successfully, until the following warnings and errors appear and the build fails.
```
...
Compiling C++ (in opt mode) /Users/me/moose/modules/heat_conduction/build/unity_src/actions_Unity.C...
Compiling C++ (in opt mode) /Users/me/moose/modules/heat_conduction/build/unity_src/fvbcs_Unity.C...
Compiling C++ (in opt mode) /Users/me/moose/modules/heat_conduction/src/base/HeatConductionApp.C...
ld: warning: -pie being ignored. It is only used when linking a main executable
ld: warning: -pie being ignored. It is only used when linking a main executable
ld: warning: -pie being ignored. It is only used when linking a main executable
ld: warning: -pie being ignored. It is only used when linking a main executable
ld: warning: -pie being ignored. It is only used when linking a main executable
Compiling Fortan Plugin (in opt mode) /Users/me/moose/modules/tensor_mechanics/test/plugins/elastic_incremental.f...
x86_64-apple-darwin13.4.0-gfortran: warning: could not understand version 12.0
Compiling Fortan Plugin (in opt mode) /Users/me/moose/modules/tensor_mechanics/test/plugins/elastic_print_multiple_fields.f...
ld: library not found for -lgcc_s.10.4
collect2: error: ld returned 1 exit status
make: *** [/Users/me/moose/framework/build.mk:325: /Users/me/moose/modules/tensor_mechanics/test/plugins/elastic_incremental-opt.plugin] Error 1
make: *** Waiting for unfinished jobs....
x86_64-apple-darwin13.4.0-gfortran: warning: could not understand version 12.0
ld: library not found for -lgcc_s.10.4
collect2: error: ld returned 1 exit status
make: *** [/Users/me/moose/framework/build.mk:325: /Users/me/moose/modules/tensor_mechanics/test/plugins/elastic_print_multiple_fields-opt.plugin] Error 1
ld: warning: -pie being ignored. It is only used when linking a main executable
ld: warning: -pie being ignored. It is only used when linking a main executable
ld: warning: -pie being ignored. It is only used when linking a main executable
```

I've got an Intel processor Macbook Pro with Ventura 13.5.

**Comment by GiudGiud** [comment]
Hello

Could you please run the diagnostics script in moose/scripts?

Something s going wrong with the compilers. Usually it's Xcode s fault but let s check

Guillaume

**Reply by GiudGiud** [reply]
@milljm have you seen this error before?
@dschwen any idea what could be bringing in gcc_s for fortran plugins?

**Reply by milljm** [reply]
```pre
AR=x86_64-apple-darwin13.4.0-ar
AS=x86_64-apple-darwin13.4.0-as
BUILD=x86_64-apple-darwin13.4.0
```

This suggests "Big Sur". So I think @GiudGiud's assumption of Xcode is correct. When you upgraded your MacOS version to Ventura, Mac _fails_ to also upgrade Xcode in the process. They leave that to you (and don't tell you!). Please attempt to upgrade your Xcode (or Xcode Command Line Tools). I believe the following command may get that started:

```bash
xcode-select --install
```
If you run into the following:

```bash
❯ xcode-select --install
xcode-select: error: command line tools are already installed, use "Software Update" in System Settings to install updates
```

You'll have to use the App Store to download Xcode, in full, and proceed from there. If you do end up going this route, when the download/install is complete, be sure to launch Xcode to actually _finish the installation_. Heh. Xcode is so picky.

**Reply by ln53** [reply]
Thanks for the help. After `xcode-select install` I did indeed get that error, so downloaded XCode 15.0. When I opened it, it did appear to do some additional setup. However I tried `make clobberall` and `make -j8` and am still getting the same error.

There doesn't seem to be any meaningful change to the `./diagnostics` output either, especially the `AR`, `AS`, and `BUILD` variables.

If it helps, running `gfortran --version` within the `moose` mamba environment gives the warning `gfortran: warning: could not understand version 12.0`, whereas running the same in a fresh terminal didn't cause a problem.

**Reply by GiudGiud** [reply]
~~It s probably a different gfortran in the fresh terminal and the moose environment, and that's ok.~~
EDIT: see below

I would first try to update mamba. Then if it doesnt work i d switch to the install from source way to avoid losing more time

**Reply by milljm** [reply]
You should not have `gfortran` available... Using Conda, and our packages, `gfortran` is actually supplied as `arm64-apple-darwin20.0.0-gfortran`

You can verify this by running:

```bash
mpif90 -show
```

So the fact that you do have a `gfortran` binary of some kind, can you perform a which on it?

```bash
which gfortran
```

I am thinking that you might have a Homebrew install (possibly) getting in the way of your Conda environment...

**Reply by milljm** [reply]
This is also interesting to me:

```pre
SDKROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
```
I didn't notice it before... I am not sure what sets this. But its not Conda (not that I know of). I have seen another post where this was set, and we (well I) thought it might point at part of the issue.

**Reply by ln53** [reply]
I did a completely fresh install of mamba today, tried `mamba update --all` just in case and got `All requested packages already installed`. Will attempt to install from source as suggested later in the week if this can't be solved.

Interestingly, `mpif90 -show` didn't actually show `arm64-apple-darwin20.0.0-gfortran` - I got:
```
x86_64-apple-darwin13.4.0-gfortran -I/Users/me/mambaforge3/envs/moose/include -L/Users/me/mambaforge3/envs/moose/lib -Wl,-rpath,/Users/me/mambaforge3/envs/moose/lib -Wl,-commons,use_dylibs -I/Users/me/mambaforge3/envs/moose/include -I/Users/me/mambaforge3/envs/moose/include -L/Users/me/mambaforge3/envs/moose/lib -lmpifort -lmpi -lpmpi
```
And `which gfortran` gives me:
```
/Users/me/mambaforge3/envs/moose/bin/gfortran
```

**Reply by milljm** [reply]
Huh, maybe this is some sort of difference between Arm and Intel Macs?

I don't mean this part:
`arm64-apple-darwin20.0.0-gfortran` and `x86_64-apple-darwin13.4.0-gfortran`. That most certainly will be the difference between Arm and Intel.

But the part about Intel having an actual `gfortran` listing... I don't have that on my machine. @GiudGiud  do you?

**Reply by GiudGiud** [reply]
Intel machine is at home rn. I can check late tonight

**Reply by milljm** [reply]
Using one of our test machines: indeed. It's an Intel thing.

```bash
(testing) $ which gfortran
/Users/civet0/projects/mamba3/envs/testing/bin/gfortran

(testing) $ gfortran --version
GNU Fortran (GCC) 9.5.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

I also see an SDKROOT being set. So that must be normal.

**Reply by milljm** [reply]
Looks like linking to gcc_c is normal.

```pre
otool -L combined-opt
        <trimmed>
	@rpath/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)
```
So where does this 10.4 version come from?

Our Mac Intel `moose-dev` package has the following gfortran dependencies:
```pre
 gfortran_impl_osx-64      9.5.0
 gfortran_osx-64           9.5.0 
```

Our Linux `moose-dev` package has teh following gfortran dependencies:
```pre
gfortran_impl_linux-64    10.4.0
gfortran_linux-64         10.4.0
```

The only thing I can think of; is this some sort of "I copied moose from my other linux computer to my machine?" thing...

**Reply by ln53** [reply]
> The only thing I can think of; is this some sort of "I copied moose from my other linux computer to my machine?" thing...

I've only ever used MOOSE (locally) on my MacBook, and I did a clean install of Conda and followed the [troubleshooting](https://mooseframework.inl.gov/help/troubleshooting.html#buildissues) guide to clean out the MOOSE repo. I had also been able to build MOOSE on this laptop a while back (probably at least a year now).

Thanks both for looking into this, I'm happy to try installing from source as suggested, as there doesn't seem to be an immediately obvious solution for now.

**Reply by milljm** [reply]
Do you have any `*.plugin` files located in `/Users/me/moose/modules/tensor_mechanics/test/plugins` ?
If so, can you perform an `otool -L` on it, and report the results?

My machine, as an example:

```bash
cd /Users/me/projects/moose/modules/tensor_mechanics/test/plugins
ls | grep plugin
otool -L elastic_dpredef-opt.plugin
```

**Reply by ln53** [reply]
I do - `otool -L elastic2-opt.plugin` gives:
```
elastic2-opt.plugin:
        /Users/me/moose/modules/tensor_mechanics/test/plugins/elastic2-opt.plugin (compatibility version 0.0.0, current version 0.0.0)
        @rpath/libmpicxx.12.dylib (compatibility version 15.0.0, current version 15.2.0)
        @rpath/libmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
        @rpath/libpmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
        @rpath/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
        @rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.0.0)
```

**Reply by milljm** [reply, marked as answer]
I imagine the only ones that are there, probably are not linking to gcc... I was hoping to see _something_ that might explain a stray path. The results you posted look just fine. I am just not seeing anything out of the ordinary.

Lets try with a very vanilla environment. I want to eliminate any possible environment variables that might be causing this issue.

```bash
env -i bash --norc --noprofile
export HOME=/Users/me
export PATH=$HOME/mambaforge3/bin:$PATH
source activate $HOME/mambaforge3/envs/moose
cd $HOME/moose
git clean -xfd    # save any uncommitted work before running
git submodule foreach --recursive git clean -xfd # save any uncommitted work before running
cd modules/tensor_mechanics
make -j 12
```
Be sure to change `me` to your actual user ID.

Edit: Sorry I kept changing some commands for you to run... I am finished now!

**Reply by ln53** [reply]
Ok thanks, I gave that a go and it built fine - still a couple of `ld: warning: -pie being ignored. It is only used when linking a main executable` warnings but they didn't cause a problem. Have just set off a build for my app in the same environment, fingers crossed!

**Reply by milljm** [reply]
You can safely ignore those ld warnings.

Well if it made it this far... Looks like we need to investigate everything that is in your PATH:

```pre
PATH=/Users/me/mambaforge3/envs/moose/bin:/Users/me/mambaforge3/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/me/bin:/usr/local/munki:/opt/X11/bin:/Users/me/mambaforge3/condabin:/Users/me/mambaforge3/bin:/Users/me/mambaforge3/envs/moose/wasp/bin
```
Somewhere among these paths, there might exist some conflicting binaries or whatnot... The only one that stands out to me is `/usr/local/munki`. Or perhaps if you've been installing other things to `/Users/me/bin` ? Regardless... something in the normal running environment is causing the issue!

**Reply by milljm** [reply]
Just read about `munki`. That probably isn't the issue. Truly, I am not sure what is hindering your builds according to the provided diagnostic log...

**Reply by ln53** [reply]
Okay I'll look into it a bit further and have a play around - worst comes to worst the "vanilla" method you describe is working. Will update here if I find the issue - thanks again.

**Reply by milljm** [reply]
Well I am glad you're satisfied using the vanilla environment method! Its a neat trick I like to use as a last resort, because if it does work, it usually means something really strange is happening in your start up profiles...

Scrutinize your `~/.zshrc` carefully (or `~/.bashrc`, `~/.bash_profile`)! is the only advice I can give...

**Reply by ln53** [reply]
Found an `export MACOSX_DEPLOYMENT_TARGET=12.0` in my `~/.zshrc` (no recollection of how or when that was added...), which suspiciously matched the `could not understand version 12.0` errors I was getting. Sure enough, removing it solved the issue - obvious in hindsight!

I'll leave your comment above marked as the answer, as it's a more general fix (or starting point at least) for this sort of problem. Really appreciate the help with this one, thanks.

**Reply by milljm** [reply]
Hah! Nice. I saw that in the diagnostic log, didn't think much of it. We use something similar in our Conda recipes. I always thought it was fairly benign. Nice to be proven wrong!

Thank you for reporting this!
